File: polars-explorer/src/Typing.ts
import {Channel} from "@tauri-apps/api/core";

export interface ColumnInfo {
    datatype: string,
    name: string,
}

// Data for individual columns
export interface ColumnData extends ColumnInfo {
    values: [],
}

export type Schema = ColumnInfo[];

// Data to be displayed in the table
export type TableData = ColumnData[];

// JSON load from the backend
export type DataJSON = {
    columns: ColumnData[]
}

export type DataChannel = Channel<DataJSON>;

// Key associated with each DataFrame
export type Key = number;

export type DataFrameInfo = {
    key: Key,
    name: string,
    length: number,
}

export type QueryInfo = {
    plan: string
}

export type DataViewInfo = {
    name: string,
    key: number,
    rows: number,
    cols: number,
    queryInfo: QueryInfo
}

export type DataInfo = {
    frameInfo: DataFrameInfo,
    viewInfo: DataViewInfo
}

export type InfoChannel = Channel<DataInfo>;

// In order to work with Immer, we need to use vanilla object rather than map
// Which is serializable
//export type DataFrameMap = Map<Key, DataFrameInfo>;
export type DataFrameMap = {
    [frameKey: Key]: DataFrameInfo
}

export type ViewMap = {
    [viewKey: Key]: DataViewInfo
}

// Map frame to their views
export type FrameViewMap = {
    [frameKey: Key]: ViewMap
}


export type FrameViewKey = [frameKey: number, viewKey: number];

export type PageInfo = {
    pageSize: number,
    currentPage: number,
    totalPage: number,
}

export type PageChannel = Channel<PageInfo>;

// This type is used to pass an array from the frontend to the backend
// It should be a serialized Array in JSON format
export type JSONArray = string;

export type ClearChannel = Channel<boolean>;

File: polars-explorer/src-tauri/src/typing.rs
use std::sync::atomic::AtomicUsize;
use serde::Serialize;
use tauri::ipc::Channel;

pub type ID = AtomicUsize;
pub type JSONValue = serde_json::Value;
pub type DataChannel = Channel<JSONValue>;

#[derive(Clone, Serialize)]
pub struct PageInfo {
    pub pageSize: usize,
    pub currentPage: usize,
    pub totalPage: usize,
}
pub type PageChannel = Channel<PageInfo>;

#[derive(Clone, Serialize)]
pub struct DataFrameInfo {
    pub key: usize,
    pub name: String,
}

pub type InfoChannel = Channel<DataFrameInfo>;